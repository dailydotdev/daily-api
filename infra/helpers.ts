import * as pulumi from '@pulumi/pulumi';
import * as gcp from '@pulumi/gcp';
import * as inputs from '@pulumi/gcp/types/input';

function camelToUnderscore(key: string): string {
  const result = key.replace(/([A-Z])/g, ' $1');
  return result.split(' ').join('_').toUpperCase();
}

export const config = new pulumi.Config();
export const stack = pulumi.getStack();
export const location = gcp.config.region || 'us-central1';

export const infra = new pulumi.StackReference(`idoshamun/infra/${stack}`);

export function createEnvVarsFromSecret(
  prefix: string,
): pulumi.Input<inputs.cloudrun.ServiceTemplateSpecContainerEnv>[] {
  const envVars = config.requireObject<Record<string, string>>('env');
  return Object.keys(envVars).map(
    (key): pulumi.Input<inputs.cloudrun.ServiceTemplateSpecContainerEnv> => {
      const secret = new gcp.secretmanager.Secret(`${prefix}-secret-${key}`, {
        secretId: `${prefix}-secret-${key}`,
        replication: { automatic: true },
      });

      const version = new gcp.secretmanager.SecretVersion(
        `${prefix}-sv-${key}`,
        {
          enabled: true,
          secret: secret.name,
          secretData: envVars[key],
        },
      );
      return {
        name: camelToUnderscore(key),
        value: pulumi
          .all([secret.secretId, version.id])
          .apply(
            ([name, version]) =>
              `gcp:///${name}/${version.split('/').reverse()[0]}`,
          ),
      };
    },
  );
}

export function serviceAccountToMember(
  serviceAccount: gcp.serviceaccount.Account,
): pulumi.Output<string> {
  return serviceAccount.email.apply((email) => `serviceAccount:${email}`);
}

export type IAMRole = { name: string; role: string };

export function addIAMRolesToServiceAccount(
  prefix: string,
  roles: IAMRole[],
  serviceAccount: gcp.serviceaccount.Account,
): void {
  const member = serviceAccountToMember(serviceAccount);
  roles.forEach(
    (role) =>
      new gcp.projects.IAMMember(`${prefix}-iam-${role.name}`, {
        role: role.role,
        member,
      }),
  );
}
