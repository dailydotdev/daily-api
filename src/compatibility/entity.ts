import { SourceRequest } from '../entity';
import { ChangeObject } from '../types';

export interface LegacySourceRequest {
  id: string;
  url: string;
  userId: string;
  userName?: string;
  userEmail?: string;
  approved?: boolean;
  closed?: boolean;
  pubId?: string;
  pubName?: string;
  pubImage?: string;
  pubTwitter?: string;
  pubRss?: string;
  createdAt: Date;
  reason?: string;
}

export const toLegacySourceRequest = (
  req: ChangeObject<SourceRequest> | SourceRequest,
): LegacySourceRequest => ({
  id: req.id,
  url: req.sourceUrl,
  userId: req.userId,
  userName: req.userName,
  userEmail: req.userEmail,
  approved: req.approved,
  closed: req.closed,
  pubId: req.sourceId,
  pubName: req.sourceName,
  pubImage: req.sourceImage,
  pubTwitter: req.sourceTwitter,
  pubRss: req.sourceFeed,
  createdAt: new Date(req.createdAt),
  reason: req.reason,
});
