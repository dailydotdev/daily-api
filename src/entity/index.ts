export * from './user';
export * from './AdvancedSettings';
export * from './Alerts';
export * from './Banner';
export * from './Bookmark';
export * from './BookmarkList';
export * from './DevCard';
export * from './Feed';
export * from './FeedAdvancedSettings';
export * from './FeedSource';
export * from './FeedTag';
export * from './Integration';
export * from './posts';
export * from './PostReport';
export * from './PostTag';
export * from './Settings';
export * from './Source';
export * from './SourceDisplay';
export * from './SourceFeed';
export * from './SourceMember';
export * from './SourceRequest';
export * from './TagSegment';
export * from './View';
export * from './Comment';
export * from './Keyword';
export * from './PostKeyword';
export * from './CommentMention';
export * from './ReputationEvent';
export * from './Submission';
export * from './notifications';
export * from './Feature';
export * from './Invite';
export * from './ContentImage';
export * from './MarketingCta';
export * from './SquadPublicRequest';
