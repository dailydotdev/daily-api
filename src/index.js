import Koa from 'koa';
import bodyParser from 'koa-bodyparser';
import KoaPinoLogger from 'koa-pino-logger';
import Router from 'koa-router';
import etag from 'koa-etag';


import config from './config';
// import compress from './middlewares/compress';
import errorHandler from './middlewares/errorHandler';
import logger from './logger';

import health from './routes/health';
import sources from './routes/sources';
import posts from './routes/posts';
import publications from './routes/publications';
import tweet from './routes/tweet';
import settings from './routes/settings';
import feeds from './routes/feeds';
import notifications from './routes/notifications';
import tags from './routes/tags';

const app = new Koa();

app.keys = [config.cookies.secret];

app.proxy = true;

// TODO: disabled due to performance issues
// app.use(compress());

app.use(bodyParser());
app.use(KoaPinoLogger({ logger, useLevel: 'debug' }));
app.use(errorHandler());
app.use(etag());

// Machine-to-machine authentication
app.use((ctx, next) => {
  if (ctx.request.get('authorization') === `Service ${config.accessSecret}`
    && ctx.request.get('user-id') && ctx.request.get('logged-in')) {
    // eslint-disable-next-line
    ctx.state = {
      user: {
        userId: ctx.request.get('user-id'),
      },
      service: true,
    };
  } else {
    delete ctx.request.headers['user-id'];
    delete ctx.request.headers['logged-in'];
  }
  return next();
});

const router = new Router({
  prefix: '/v1',
});

router.use(feeds.routes(), feeds.allowedMethods());
router.use(sources.routes(), sources.allowedMethods());
router.use(posts.routes(), posts.allowedMethods());
router.use(publications.routes(), publications.allowedMethods());
router.use(tweet.routes(), tweet.allowedMethods());
router.use(settings.routes(), settings.allowedMethods());
router.use(notifications.routes(), notifications.allowedMethods());
router.use(tags.routes(), tags.allowedMethods());

app.use(router.routes(), router.allowedMethods());
app.use(health.routes(), health.allowedMethods());

export default app;
