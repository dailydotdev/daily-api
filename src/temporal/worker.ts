import { NativeConnection } from '@temporalio/worker';
import { getTemporalServerOptions } from './config';

let connection: NativeConnection;

export const getTemporalWorkerConnection = async () => {
  if (connection) {
    return connection;
  }

  const { tls, address } = getTemporalServerOptions();

  connection = await NativeConnection.connect({ address, tls });

  return connection;
};
