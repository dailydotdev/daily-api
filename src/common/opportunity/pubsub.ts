import { DataSource, type EntityManager } from 'typeorm';
import { FastifyBaseLogger } from 'fastify';
import {
  CandidateAcceptedOpportunityMessage,
  CandidatePreferenceUpdated,
  MatchedCandidate,
  OpportunityMessage,
  RecruiterAcceptedCandidateMatchMessage,
  Salary,
  UserCV,
} from '@dailydotdev/schema';
import { demoCompany, triggerTypedEvent } from '../../common';
import { getSecondsTimestamp } from '../date';
import { UserCandidatePreference } from '../../entity/user/UserCandidatePreference';
import { ChangeObject } from '../../types';
import { OpportunityMatch } from '../../entity/OpportunityMatch';
import { OpportunityJob } from '../../entity/opportunities/OpportunityJob';
import { UserCandidateKeyword } from '../../entity/user/UserCandidateKeyword';
import { queryReadReplica } from '../queryReadReplica';
import { ContentPreference } from '../../entity/contentPreference/ContentPreference';
import {
  ContentPreferenceStatus,
  ContentPreferenceType,
} from '../../entity/contentPreference/types';
import { ContentPreferenceOrganization } from '../../entity/contentPreference/ContentPreferenceOrganization';

const fetchCandidateKeywords = async (
  manager: EntityManager,
  candidatePreference: UserCandidatePreference | null,
): Promise<Array<string>> => {
  if (!candidatePreference) {
    return [];
  }

  // Fetch custom keywords if enabled
  if (candidatePreference.customKeywords) {
    const customKeywords = await manager
      .getRepository(UserCandidateKeyword)
      .findBy({
        userId: candidatePreference.userId,
      });

    return customKeywords.map((k) => k.keyword);
  }

  // Otherwise fetch keywords from content preferences
  const feedKeywords = await manager
    .createQueryBuilder()
    .select(`"keywordId"`, 'keyword')
    .from(ContentPreference, 'cpk')
    .where(`cpk."feedId" = :userId`, {
      userId: candidatePreference.userId,
    })
    .andWhere('cpk.type = :contentPreferenceType', {
      contentPreferenceType: ContentPreferenceType.Keyword,
    })
    .andWhere('cpk.status != :contentPreferenceStatus', {
      contentPreferenceStatus: ContentPreferenceStatus.Blocked,
    })
    .getRawMany<{ keyword: string }>();

  return feedKeywords.map((k) => k.keyword);
};

export const notifyOpportunityMatchAccepted = async ({
  con,
  logger,
  data,
}: {
  con: DataSource;
  logger: FastifyBaseLogger;
  data: ChangeObject<OpportunityMatch> | null;
}) => {
  if (!data) {
    logger.warn('No data provided for opportunity match accepted notification');
    return;
  }

  const { match, candidatePreference, keywords } = await queryReadReplica(
    con,
    async ({ queryRunner }) => {
      const [match, candidatePreference] = await Promise.all([
        queryRunner.manager.getRepository(OpportunityMatch).findOneBy({
          opportunityId: data.opportunityId,
          userId: data.userId,
        }),
        queryRunner.manager
          .getRepository(UserCandidatePreference)
          .findOneBy({ userId: data.userId }),
      ]);

      const keywords = await fetchCandidateKeywords(
        queryRunner.manager,
        candidatePreference,
      );

      return { match, candidatePreference, keywords };
    },
  );

  if (!match) {
    logger.warn(
      { opportunityId: data.opportunityId, userId: data.userId },
      'Opportunity match not found for accepted notification',
    );
    return;
  }

  if (!candidatePreference) {
    logger.warn(
      { userId: data.userId },
      'Candidate preference not found for user accepting opportunity',
    );
    return;
  }

  const message = new CandidateAcceptedOpportunityMessage({
    opportunityId: match.opportunityId,
    userId: match.userId,
    createdAt: getSecondsTimestamp(match.createdAt),
    updatedAt: getSecondsTimestamp(match.updatedAt),
    screening: match.screening,
    candidatePreference: {
      ...candidatePreference,
      salaryExpectation: new Salary({
        min: candidatePreference.salaryExpectation?.min
          ? BigInt(candidatePreference.salaryExpectation.min)
          : undefined,
        period: candidatePreference.salaryExpectation?.period ?? undefined,
      }),
      cv: new UserCV({
        ...candidatePreference.cv,
        lastModified:
          getSecondsTimestamp(candidatePreference.cv.lastModified || 0) ||
          undefined,
      }),
      updatedAt: getSecondsTimestamp(candidatePreference.updatedAt),
      keywords: keywords,
    },
  });

  try {
    await triggerTypedEvent(
      logger,
      'api.v1.candidate-accepted-opportunity',
      message,
    );
  } catch (_err) {
    const err = _err as Error;
    logger.error(
      { err, message },
      'failed to send opportunity match accepted event',
    );
  }
};

export const notifyRecruiterCandidateMatchAccepted = async ({
  con,
  logger,
  data,
}: {
  con: DataSource;
  logger: FastifyBaseLogger;
  data: ChangeObject<OpportunityMatch> | null;
}) => {
  if (!data) {
    logger.warn(
      'No data provided for candidate opportunity match accepted notification',
    );
    return;
  }

  const match = await queryReadReplica(con, async ({ queryRunner }) => {
    return queryRunner.manager.getRepository(OpportunityMatch).findOne({
      select: ['opportunityId', 'userId'],
      where: { opportunityId: data.opportunityId, userId: data.userId },
    });
  });

  if (!match) {
    logger.warn(
      { opportunityId: data.opportunityId, userId: data.userId },
      'Opportunity match not found for recruiter accepted candidate notification',
    );
    return;
  }

  const message = new RecruiterAcceptedCandidateMatchMessage({
    opportunityId: match.opportunityId,
    userId: match.userId,
    createdAt: getSecondsTimestamp(match.createdAt),
    updatedAt: getSecondsTimestamp(match.updatedAt),
  });

  try {
    await triggerTypedEvent(
      logger,
      'api.v1.recruiter-accepted-candidate-match',
      message,
    );
  } catch (_err) {
    const err = _err as Error;
    logger.error(
      { err, message },
      'failed to send recruiter accepted candidate match event',
    );
  }
};

export const notifyJobOpportunity = async ({
  con,
  logger,
  isUpdate,
  opportunityId,
}: {
  con: DataSource;
  logger: FastifyBaseLogger;
  isUpdate: boolean;
  opportunityId: string;
}) => {
  const topicName = isUpdate
    ? 'api.v1.opportunity-updated'
    : 'api.v1.opportunity-added';

  const [opportunity, organization, keywords] = await queryReadReplica(
    con,
    async ({ queryRunner }) => {
      const opportunity = await queryRunner.manager
        .getRepository(OpportunityJob)
        .findOneOrFail({
          where: { id: opportunityId },
          relations: {
            organization: true,
            keywords: true,
          },
        });

      const [organization, keywords] = await Promise.all([
        opportunity.organization,
        opportunity.keywords,
      ]);

      return [opportunity, organization, keywords];
    },
  );

  if (!organization) {
    logger.warn(
      {
        opportunityId: opportunity.id,
        organizationId: opportunity.organizationId,
      },
      'opportunity has no organization, skipping',
    );
    return;
  }

  /**
   * Demo logic: if the company is the demo company we can omit using Gondul and simply return the users from that company as matched candidates
   */
  if (organization.id === demoCompany.id) {
    const members = await con
      .getRepository(ContentPreferenceOrganization)
      .find({
        select: ['userId'],
        where: { organizationId: organization.id },
      });
    for (const { userId } of members) {
      await triggerTypedEvent(
        logger,
        'gondul.v1.candidate-opportunity-match',
        new MatchedCandidate({
          opportunityId,
          userId,
          matchScore: 0.87,
          reasoning:
            "We have noticed that you've been digging into React performance optimization and exploring payment systems lately. Your skills in TypeScript and Node.js line up directly with the core technologies this team uses. You also follow several Atlassian engineers and have shown consistent interest in project management software, which makes this role a natural fit for your trajectory.",
          reasoningShort:
            'Your skills in TypeScript and Node.js line up directly with the core technologies this team uses.',
        }),
      );
    }
    return;
  }

  const message = new OpportunityMessage({
    opportunity: {
      ...opportunity,
      createdAt: getSecondsTimestamp(opportunity.createdAt),
      updatedAt: getSecondsTimestamp(opportunity.updatedAt),
      keywords: keywords.map((k) => k.keyword),
    },
    organization: {
      ...organization,
      createdAt: getSecondsTimestamp(organization.createdAt),
      updatedAt: getSecondsTimestamp(organization.updatedAt),
    },
  });

  try {
    await triggerTypedEvent(logger, topicName, message);
  } catch (_err) {
    const err = _err as Error;
    logger.error({ err, message }, 'failed to send opportunity event');
  }
};

export const notifyCandidatePreferenceChange = async ({
  con,
  logger,
  userId,
}: {
  con: DataSource;
  logger: FastifyBaseLogger;
  userId: string;
}) => {
  const { candidatePreference, keywords } = await queryReadReplica(
    con,
    async ({ queryRunner }) => {
      const candidatePreference = await queryRunner.manager
        .getRepository(UserCandidatePreference)
        .findOneBy({ userId: userId });

      const keywords = await fetchCandidateKeywords(
        queryRunner.manager,
        candidatePreference,
      );

      return { candidatePreference, keywords };
    },
  );

  if (!candidatePreference) {
    logger.warn(
      { userId },
      'Candidate preference not found for user, skipping notification',
    );
    return;
  }

  const message = new CandidatePreferenceUpdated({
    payload: {
      ...candidatePreference,
      salaryExpectation: new Salary({
        min: candidatePreference.salaryExpectation?.min
          ? BigInt(candidatePreference.salaryExpectation.min)
          : undefined,
        period: candidatePreference.salaryExpectation?.period ?? undefined,
      }),
      cv: new UserCV({
        ...candidatePreference?.cv,
        lastModified:
          getSecondsTimestamp(candidatePreference?.cv?.lastModified || 0) ||
          undefined,
      }),
      updatedAt:
        getSecondsTimestamp(candidatePreference?.updatedAt) || undefined,
      keywords: keywords,
    },
  });

  try {
    await triggerTypedEvent(
      logger,
      'api.v1.candidate-preference-updated',
      message,
    );
  } catch (_err) {
    const err = _err as Error;
    logger.error(
      { err, message },
      'failed to send candidate preference updated event',
    );
  }
};
