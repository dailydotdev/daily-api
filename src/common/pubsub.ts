import { PubSub, Topic } from '@google-cloud/pubsub';
import { FastifyBaseLogger } from 'fastify';
import {
  Post,
  Settings,
  Submission,
  User,
  CommentMention,
  SourceMember,
  Feature,
  Source,
  ArticlePost,
  PostMention,
  Comment,
  ContentImage,
  Banner,
  FreeformPost,
  CollectionPost,
  NotificationV2,
  UserPersonalizedDigest,
} from '../entity';
import { ChangeMessage, ChangeObject } from '../types';
import { SourceMemberRoles } from '../roles';
import {
  addPubsubSpanLabels,
  opentelemetry,
  runInRootSpan,
  runInSpan,
} from '../telemetry/opentelemetry';
import { Message } from '@google-cloud/pubsub';
// import { performance } from 'perf_hooks';
import { DataSource } from 'typeorm';
import { FastifyLoggerInstance } from 'fastify';
import pino from 'pino';
import { PersonalizedDigestFeatureConfig } from '../growthbook';

export const pubsub = new PubSub();
const postCommentedTopic = pubsub.topic('post-commented');
const commentCommentedTopic = pubsub.topic('comment-commented');
const commentFeaturedTopic = pubsub.topic('comment-featured');
const commentsUpdateTopic = pubsub.topic('update-comments');
const usernameChangedTopic = pubsub.topic('username-changed');
const settingsUpdatedTopic = pubsub.topic('settings-updated');
const sendAnalyticsReportTopic = pubsub.topic('send-analytics-report');
const viewsTopic = pubsub.topic('views');
const postBannedOrRemovedTopic = pubsub.topic('post-banned-or-removed');
const sourceFeedAddedTopic = pubsub.topic('source-feed-added');
const sourceFeedRemovedTopic = pubsub.topic('source-feed-removed');
const communityLinkAccessTopic = pubsub.topic('community-link-access');
const communityLinkRejectedTopic = pubsub.topic('community-link-rejected');
const newNotificationTopic = pubsub.topic('api.v1.new-notification');
const newPostMentionTopic = pubsub.topic('api.v1.new-post-mention');
const newCommentMentionTopic = pubsub.topic('api.v1.new-comment-mention');
const memberJoinedSourceTopic = pubsub.topic('api.v1.member-joined-source');
const featureAccess = pubsub.topic('api.v1.feature-granted');
const sourcePrivacyUpdatedTopic = pubsub.topic('api.v1.source-privacy-updated');
const featuresResetTopic = pubsub.topic('features-reset');
const contentRequestedTopic = pubsub.topic('api.v1.content-requested');
const postVisibleTopic = pubsub.topic('api.v1.post-visible');
const bannerAddedTopic = pubsub.topic('api.v1.banner-added');
const bannerRemovedTopic = pubsub.topic('api.v1.banner-deleted');
const sourceMemberRoleChangedTopic = pubsub.topic(
  'api.v1.source-member-role-changed',
);
const contentImageDeletedTopic = pubsub.topic('api.v1.content-image-deleted');
const postContentEditedTopic = pubsub.topic('api.v1.post-content-edited');
const commentEditedTopic = pubsub.topic('api.v1.comment-edited');
const commentDeletedTopic = pubsub.topic('api.v1.comment-deleted');
const sourceCreatedTopic = pubsub.topic('api.v1.source-created');
const generatePersonalizedDigestTopic = pubsub.topic(
  'api.v1.generate-personalized-digest',
);
const postYggdrasilIdSet = pubsub.topic('api.v1.post-yggdrasil-id-set');
const postCollectionUpdatedTopic = pubsub.topic(
  'api.v1.post-collection-updated',
);
const userReadmeUpdatedTopic = pubsub.topic('api.v1.user-readme-updated');
const userReputationUpdatedTopic = pubsub.topic('user-reputation-updated');

export enum NotificationReason {
  New = 'new',
  Publish = 'publish',
  Approve = 'approve',
  Decline = 'decline',
  Exists = 'exists',
}

// Need to support console as well
export type EventLogger = Omit<FastifyBaseLogger, 'fatal'>;

export const publishEvent = async (
  log: EventLogger,
  topic: Topic,
  payload: Record<string, unknown>,
): Promise<void> =>
  runInSpan(
    `publishEvent ${topic.name}`,
    async () => {
      if (
        process.env.NODE_ENV === 'production' ||
        process.env.ENABLE_PUBSUB === 'true'
      ) {
        try {
          await topic.publishMessage({
            json: payload,
          });
        } catch (err) {
          log.error(
            { err, topic: topic.name, payload },
            'failed to publish message',
          );
        }
      }
    },
    {
      kind: opentelemetry.SpanKind.PRODUCER,
    },
  );

export const notifyPostCommented = async (
  log: EventLogger,
  postId: string,
  userId: string,
  commentId: string,
  contentHtml: string,
): Promise<void> =>
  publishEvent(log, postCommentedTopic, {
    postId,
    userId,
    commentId,
    contentHtml,
  });

export const notifyCommentCommented = async (
  log: EventLogger,
  postId: string,
  userId: string,
  parentCommentId: string,
  childCommentId: string,
  contentHtml: string,
): Promise<void> =>
  publishEvent(log, commentCommentedTopic, {
    postId,
    userId,
    parentCommentId,
    childCommentId,
    contentHtml,
  });

export const notifyCommentFeatured = async (
  log: EventLogger,
  commentId: string,
): Promise<void> =>
  publishEvent(log, commentFeaturedTopic, {
    commentId,
  });

export const notifyCommentsUpdate = async (
  log: EventLogger,
  oldUsername: string,
  newUsername: string,
  commentIds: string[],
): Promise<void> =>
  publishEvent(log, commentsUpdateTopic, {
    oldUsername,
    newUsername,
    commentIds,
  });

export const notifyUserReadmeUpdated = (
  log: EventLogger,
  user: ChangeObject<User>,
): Promise<void> => publishEvent(log, userReadmeUpdatedTopic, { user });

export const notifyUsernameChanged = (
  log: EventLogger,
  userId: string,
  oldUsername: string,
  newUsername: string,
): Promise<void> =>
  publishEvent(log, usernameChangedTopic, { userId, oldUsername, newUsername });

export const notifySettingsUpdated = (
  log: EventLogger,
  settings: ChangeObject<Settings>,
): Promise<void> => publishEvent(log, settingsUpdatedTopic, settings);

export const notifySourcePrivacyUpdated = (
  log: EventLogger,
  source: ChangeObject<Source>,
): Promise<void> => publishEvent(log, sourcePrivacyUpdatedTopic, { source });

export const notifySendAnalyticsReport = async (
  log: EventLogger,
  postId: string,
): Promise<void> =>
  publishEvent(log, sendAnalyticsReportTopic, {
    postId,
  });

export const notifyView = (
  log: EventLogger,
  postId: string,
  userId: string,
  referer: string,
  timestamp: Date,
  tags?: string[],
): Promise<void> =>
  publishEvent(log, viewsTopic, {
    postId,
    userId,
    referer,
    timestamp,
    tags,
  });

export const notifyPostBannedOrRemoved = async (
  log: EventLogger,
  post: ChangeObject<Post>,
): Promise<void> =>
  publishEvent(log, postBannedOrRemovedTopic, {
    post,
  });

export const notifySourceFeedAdded = async (
  log: EventLogger,
  sourceId: string,
  feed: string,
): Promise<void> =>
  publishEvent(log, sourceFeedAddedTopic, {
    feed,
    sourceId,
  });

export const notifySourceFeedRemoved = async (
  log: EventLogger,
  sourceId: string,
  feed: string,
): Promise<void> =>
  publishEvent(log, sourceFeedRemovedTopic, {
    feed,
    sourceId,
  });

export const notifySubmissionRejected = async (
  log: EventLogger,
  submission: ChangeObject<Submission>,
): Promise<void> => publishEvent(log, communityLinkRejectedTopic, submission);

export const notifySubmissionGrantedAccess = async (
  log: EventLogger,
  userId: string,
): Promise<void> => publishEvent(log, communityLinkAccessTopic, { userId });

export const notifySourceMemberRoleChanged = async (
  log: EventLogger,
  previousRole: SourceMemberRoles,
  sourceMember: ChangeObject<SourceMember>,
): Promise<void> =>
  publishEvent(log, sourceMemberRoleChangedTopic, {
    previousRole,
    sourceMember,
  });

export const notifyNewNotification = async (
  log: EventLogger,
  notification: ChangeObject<NotificationV2>,
): Promise<void> => publishEvent(log, newNotificationTopic, { notification });

export const notifyNewPostMention = async (
  log: EventLogger,
  postMention: ChangeObject<PostMention>,
): Promise<void> => publishEvent(log, newPostMentionTopic, { postMention });

export const notifyNewCommentMention = async (
  log: EventLogger,
  commentMention: ChangeObject<CommentMention>,
): Promise<void> =>
  publishEvent(log, newCommentMentionTopic, { commentMention });

export const notifyMemberJoinedSource = async (
  log: EventLogger,
  sourceMember: ChangeObject<SourceMember>,
): Promise<void> =>
  publishEvent(log, memberJoinedSourceTopic, { sourceMember });

export const notifyFeatureAccess = async (
  log: EventLogger,
  feature: ChangeObject<Feature>,
): Promise<void> => publishEvent(log, featureAccess, { feature });

export const notifyPostVisible = async (
  log: EventLogger,
  post: ChangeObject<Post>,
): Promise<void> => publishEvent(log, postVisibleTopic, { post });

export const notifyBannerCreated = async (
  log: EventLogger,
  banner: ChangeObject<Banner>,
): Promise<void> => publishEvent(log, bannerAddedTopic, { banner });

export const notifyBannerRemoved = async (
  log: EventLogger,
  banner: ChangeObject<Banner>,
): Promise<void> => publishEvent(log, bannerRemovedTopic, { banner });

export const notifyFeaturesReset = async (log: EventLogger): Promise<void> =>
  publishEvent(log, featuresResetTopic, {});

type ContentRequestedSubmission = { submissionId: string } & Pick<
  ArticlePost,
  'sourceId' | 'url'
>;
type ContentRequestedURL = Pick<ArticlePost, 'id' | 'origin' | 'url'>;
type ContentRequestedFreeForm = Pick<
  FreeformPost,
  'id' | 'content' | 'title'
> & {
  post_type;
};
export type ContentRequested =
  | ContentRequestedSubmission
  | ContentRequestedURL
  | ContentRequestedFreeForm;

export const notifyContentRequested = async (
  log: EventLogger,
  content: ContentRequested,
): Promise<void> => publishEvent(log, contentRequestedTopic, content);

export const notifyFreeformContentRequested = async (
  logger: EventLogger,
  freeform: ChangeMessage<FreeformPost>,
): Promise<void> =>
  notifyContentRequested(logger, {
    id: freeform.payload.after.id,
    content: freeform.payload.after.content,
    title: freeform.payload.after.title,
    post_type: freeform.payload.after.type,
  });

export const notifyContentImageDeleted = async (
  log: EventLogger,
  contentImage: ChangeObject<ContentImage>,
): Promise<void> =>
  publishEvent(log, contentImageDeletedTopic, { contentImage });

export const notifyPostContentEdited = async (
  log: EventLogger,
  post: ChangeObject<Post>,
): Promise<void> => publishEvent(log, postContentEditedTopic, { post });

export const notifyCommentEdited = async (
  log: EventLogger,
  comment: ChangeObject<Comment>,
): Promise<void> => publishEvent(log, commentEditedTopic, { comment });

export const notifyCommentDeleted = async (
  log: EventLogger,
  comment: ChangeObject<Comment>,
): Promise<void> => publishEvent(log, commentDeletedTopic, { comment });

export const notifySourceCreated = async (
  log: EventLogger,
  source: ChangeObject<Source>,
): Promise<void> => publishEvent(log, sourceCreatedTopic, { source });

export const notifyGeneratePersonalizedDigest = async ({
  log,
  personalizedDigest,
  emailSendTimestamp,
  previousSendTimestamp,
  emailBatchId,
  deduplicate,
  config,
}: {
  log: EventLogger;
  personalizedDigest: UserPersonalizedDigest;
  emailSendTimestamp: number;
  previousSendTimestamp: number;
  emailBatchId?: string;
  deduplicate?: boolean;
  config?: PersonalizedDigestFeatureConfig;
}): Promise<void> =>
  publishEvent(log, generatePersonalizedDigestTopic, {
    personalizedDigest,
    emailSendTimestamp,
    previousSendTimestamp,
    emailBatchId,
    deduplicate: deduplicate ?? true,
    config,
  });

export const notifyPostYggdrasilIdSet = async (
  log: EventLogger,
  post: ChangeObject<Post>,
): Promise<void> =>
  publishEvent(log, postYggdrasilIdSet, {
    post,
  });

export const notifyPostCollectionUpdated = async (
  log: EventLogger,
  post: ChangeObject<CollectionPost>,
): Promise<void> => publishEvent(log, postCollectionUpdatedTopic, { post });

export const notifyReputationIncrease = async (
  log: EventLogger,
  user: ChangeObject<User>,
  userAfter: ChangeObject<User>,
): Promise<void> =>
  publishEvent(log, userReputationUpdatedTopic, {
    user,
    userAfter,
  });

export const workerSubscribe = (
  logger: pino.Logger,
  pubsub: PubSub,
  connection: DataSource,
  subscription: string,
  meter: opentelemetry.Meter,
  handler: (
    message: Message,
    con: DataSource,
    logger: FastifyLoggerInstance,
    pubsub: PubSub,
  ) => Promise<void>,
  maxMessages = 1,
): void => {
  logger.info(`subscribing to ${subscription}`);
  const sub = pubsub.subscription(subscription, {
    flowControl: {
      maxMessages,
    },
    batching: { maxMilliseconds: 10 },
  });
  const childLogger = logger.child({ subscription });
  // const histogram = meter.createHistogram('message_processing_time', {
  //   unit: 'ms',
  //   description: 'time to process a message',
  // });
  sub.on('message', async (message) =>
    runInRootSpan(
      `message: ${subscription}`,
      async (span) => {
        // const startTime = performance.now();
        // let success = true;
        addPubsubSpanLabels(span, subscription, message);
        try {
          await runInSpan('handler', async () =>
            handler(message, connection, childLogger, pubsub),
          );
          message.ack();
        } catch (err) {
          // success = false;
          childLogger.error(
            {
              messageId: message.id,
              data: message.data.toString('utf-8'),
              err,
            },
            'failed to process message',
          );
          message.nack();
        }
        // histogram.record(performance.now() - startTime, {
        //   subscription,
        //   success,
        // });
      },
      {
        kind: opentelemetry.SpanKind.CONSUMER,
      },
    ),
  );
};
