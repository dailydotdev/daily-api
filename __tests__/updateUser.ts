import { Connection, getConnection } from 'typeorm';
import { FastifyInstance } from 'fastify';
import { PubSub } from '@google-cloud/pubsub';

import appFunc from '../src';
import { mockMessage } from './helpers';
import worker from '../src/workers/updateUser';
import { User } from '../src/entity';

let con: Connection;
let app: FastifyInstance;

beforeAll(async () => {
  con = await getConnection();
  app = await appFunc();
  return app.ready();
});

it('should update an existing user', async () => {
  await con.getRepository(User).save({
    id: 'abc',
    name: 'ido',
    image: 'https://daily.dev/image.jpg',
  });

  const message = mockMessage({
    user: {
      id: 'abc',
      name: 'ido',
      email: 'ido@acme.com',
      image: 'https://daily.dev/image.jpg',
    },
    newProfile: {
      name: 'Ido',
    },
  });

  await worker.handler(message, con, app.log, new PubSub());
  expect(message.ack).toBeCalledTimes(1);
  const users = await con.getRepository(User).find();
  expect(users.length).toEqual(1);
  expect(users[0]).toMatchSnapshot();
});

it('should ack message if user does not exist', async () => {
  const message = mockMessage({
    user: {
      id: 'abc',
      name: 'ido',
      email: 'ido@acme.com',
      image: 'https://daily.dev/image.jpg',
    },
    newProfile: {
      name: 'Ido',
    },
  });

  await worker.handler(message, con, app.log, new PubSub());
  expect(message.ack).toBeCalledTimes(1);
  const users = await con.getRepository(User).find();
  expect(users.length).toEqual(0);
});
