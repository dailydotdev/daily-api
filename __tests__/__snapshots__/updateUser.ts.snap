// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should update an existing user 1`] = `
User {
  "id": "abc",
  "image": "https://daily.dev/image.jpg",
  "name": "Ido",
}
`;
